src/nms_adpcm.c

```diff
@@ -0,0 +1,1118 @@
+static unsigned int table_step [] =
+{	0x73F,	0,		0,		0,		0x1829,	0,		0,		0,		/* 2-bit */
+	0x3EB,	0,		0xC18,	0,		0x1581,	0,		0x226E,	0,		/* 3-bit */
+	0x20C,	0x635,	0xA83,	0xF12,	0x1418,	0x19E3,	0x211A,	0x2BBA,	/* 4-bit */
+} ;
+
+/* Divide-and-conquer lookup table for quantizing using table_step. */
+static int table_step_search [] =
+{	0,		0x1F6D,	0,		-0x1F6D,	0,		0,			0,			0, /* 2-bit */
+	0x1008,	0x1192,	0,		-0x219A,	0x1656,	-0x1656,	0,			0, /* 3-bit */
+	0x872,	0x1277,	-0x8E6,	-0x232B,	0xD06,	-0x17D7,	-0x11D3,	0, /* 4-bit */
+} ;
+
+
+/*============================================================================================
+** Static functions.
+*/
+
+static void nms_adpcm_update (struct nms_adpcm_state *s) ;
+static void nms_adpcm_reset_codec (struct nms_adpcm_state *s) ;
+
+static int16_t nms_adpcm_reconstruct_sample (struct nms_adpcm_state *s, uint8_t I) ;
+static uint8_t nms_adpcm_encode_sample (struct nms_adpcm_state *s, int16_t sl) ;
+static int16_t nms_adpcm_decode_sample (struct nms_adpcm_state *s, uint8_t code) ;
+
+static void nms_adpcm_block_pack_16 (const int16_t codewords [], uint8_t block [], int16_t rms) ;
+static void nms_adpcm_block_pack_24 (const int16_t codewords [], uint8_t block [], int16_t rms) ;
+static void nms_adpcm_block_pack_32 (const int16_t codewords [], uint8_t block [], int16_t rms) ;
+
+static void nms_adpcm_block_unpack_16 (uint8_t block [], int16_t codewords [], int16_t *rms) ;
+static void nms_adpcm_block_unpack_24 (uint8_t block [], int16_t codewords [], int16_t *rms) ;
+static void nms_adpcm_block_unpack_32 (uint8_t block [], int16_t codewords [], int16_t *rms) ;
+
+static int nms_adpcm_decode_block (NMS_ADPCM_PRIVATE *pnms, uint8_t block [], int16_t samples []) ;
+static int nms_adpcm_encode_block (NMS_ADPCM_PRIVATE *pnms, int16_t samples [], uint8_t block []) ;
+
+static sf_count_t nms_adpcm_read_s (SF_PRIVATE *psf, short *ptr, sf_count_t len) ;
+static sf_count_t nms_adpcm_read_i (SF_PRIVATE *psf, int *ptr, sf_count_t len) ;
+static sf_count_t nms_adpcm_read_f (SF_PRIVATE *psf, float *ptr, sf_count_t len) ;
+static sf_count_t nms_adpcm_read_d (SF_PRIVATE *psf, double *ptr, sf_count_t len) ;
+
+static sf_count_t nms_adpcm_write_s (SF_PRIVATE *psf, const short *ptr, sf_count_t len) ;
+static sf_count_t nms_adpcm_write_i (SF_PRIVATE *psf, const int *ptr, sf_count_t len) ;
+static sf_count_t nms_adpcm_write_f (SF_PRIVATE *psf, const float *ptr, sf_count_t len) ;
+static sf_count_t nms_adpcm_write_d (SF_PRIVATE *psf, const double *ptr, sf_count_t len) ;
+
+static int nms_adpcm_close (SF_PRIVATE *psf) ;
+static sf_count_t nms_adpcm_seek (SF_PRIVATE *psf, int mode, sf_count_t offset) ;
+
+/*
+** An exponential function (antilog) approximation.
+**
+** Maps [1,20480] to [1,1024] in an exponential relationship. This is
+** approximately ret = b^exp where b = e^(ln(1024)/ln(20480)) ~= 1.0003385
+*/
+static inline int antilog (int exp)
+{	int ret ;
```

Comment:
Please define functions as:
```
static inline in
antilog (int exp)
```
This makes it easier to find a functon implementation easier to find by grepping for `^function_name`.